### DISCUSSION TITLE
change watch_tick_animation in simple_clock

### DISCUSSION THREAD

**Discussion by rienesl** [joeycastillo/Sensor-Watch#155] [Ideas]
I got my new toy yesterday, and unfortunately my programming knowledge is not sufficient to even get a colon blinking. Here is my problem: The animation of the simple clock is beyond my imagination. Why are 2 digits alternating, while the colon between the hours and the minutes stay on. So I wanted to change the strange animation to a blinking colon - like every other watch does. So I had a look at the code and found the following line, responsible for that:
`if (!watch_tick_animation_is_running()) watch_start_tick_animation(500);`
So far so good, but I can not find the definition of the function "watch_start_tick_animation" anywhere. I assume it is somewhere in here:
`#include <stdlib.h>`
`#include "simple_clock_face.h"`
`#include "watch.h"`
`#include "watch_utility.h"`
But I can not find stdlib.h and in the others there is no watch_start_tick_animation defined. Can anyone help me out? Yes, I am a complete newbie in here, I have some programming knowledge in QBasic (don't laugh!), html (native, no WYSIWYG editor) and javascript, so please be a bit patient...

**Comment by rienesl** [comment]
Found it in /Sensor-Watch/watch-library/hardware/watch/watch_slcd.c:

void watch_start_tick_animation(uint32_t duration) {
    watch_display_character(' ', 8);
    const uint32_t segs[] = { SLCD_SEGID(0, 2)};
    slcd_sync_start_animation(&SEGMENT_LCD_0, segs, 1, duration);
}

Now my new problem is, that https://www.sensorwatch.net/docs/wig/display/ does not go into detail, how the colon works. What i can read out of the code is, that currently segment 8 is somehow get alternating feed with segment 0 or 2. Unfortunately I am stuck here now, until I find out, how to address the colon, so I can change the address (which pints currently to segment 8), as well as the data sent (which is currently 0 and 2 and I guess I need to change that to 0 and 1?).
Yes, I have no real clue what I am doing here at all, I know...

**Comment by joeycastillo** [comment]
You’re not lost; the issue is that the SAM L22 can only do low-power autonomous blinking on two segment lines, and autonomous segment animation on two others. Unfortunately, because of the way the F-91W’s LCD is laid out, the colon isn’t on one of those lines. As a result you need to use CPU time to blink the colon, whereas we can do our little tick/tock animation while keeping the CPU entirely asleep, saving power. 

We were lucky that these low power animation segment lines lined up with the seconds display, since updating the seconds also requires CPU time, making the seconds display less relevant in low energy mode. I admit that it’s not as traditional as a blinking colon, but my hope is that it conveys the sense that (1) the watch has entered a different mode where (2) seconds are unavailable, and (3) action is required — pressing the Alarm/Wake button — to get the seconds back.

**Reply by rienesl** [reply]
Thanks for the very detailed answer, I really appreciate it. But the answer has put up new questions:
You write, only 2 segments can blink and 2 others can do an animation.
1) could an other setup of the pins (board layout) solve that "issue" in the future?
2) what 2 other segments that you mentioned can blink?
3) the "autonomous segment animation on two others" - does it has to be an animation? basically I aim for a "blinking cursor", which I would be fine to me as well (but I am afraid I already know the answer...)
Just to clarify: None of this is dramatic, just a nice to have. If 3) is not possible, I will simply deactivate the animation - but this is simply a personal thing

**Reply by joeycastillo** [reply]
> You write, only 2 segments can blink and 2 others can do an animation.

Technically two segment lines, and since this LCD has three commons, it's 2x3 or six segments that can blink, and six that can animate. SEG0 and SEG1 allow for blinking, and SEG2 and SEG3 allow for a very basic bit shifting animation. 

> could an other setup of the pins (board layout) solve that "issue" in the future?

It's possible! Challenging but possible. The key would be to connect either the SEG0 or SEG1 line (highlighted at far right) to the SEG16 pad in the LCD strip (highlighted at the left). Since the colon is connected to SEG16, this would allow you to blink the colon in low energy mode.

<img width="1318" alt="Screen Shot 2022-12-28 at 12 46 14 AM" src="https://user-images.githubusercontent.com/1765152/209770224-2669e7d6-a2f0-4a5f-9397-902b251ab86c.png">


> what 2 other segments that you mentioned can blink?

With our board layout, these are the segments with special capabilities (green=blink, red=animate): 

![FAoi5mSXIAAxvvO](https://user-images.githubusercontent.com/1765152/209769504-1a202432-be88-4945-8fb5-194cf77221c3.png)

> the "autonomous segment animation on two others" - does it has to be an animation? basically I aim for a "blinking cursor", which I would be fine to me as well (but I am afraid I already know the answer...)

For our LCD, the animation segments aren't terribly useful since the segments aren't ordered very usefully; this shift register based animation feature is really aimed at someone designing a custom LCD with segments laid out to make use of it, like a spinner. 

<img width="825" alt="Screen Shot 2022-12-28 at 12 50 49 AM" src="https://user-images.githubusercontent.com/1765152/209770753-fdda3cb8-81d9-487d-ac6f-370ba995d54e.png">

In my experience the smallest animation you can do with those red ones involves two segments, and the first two in our segment mapping are those two in the bottom left. You can play with passing different sizes into `slcd_sync_start_animation`, but I get the sense that what we have now is what we can get out of that feature.

**Reply by rienesl** [reply]
rienesl
Dec 29, 2022
Author

Whow, again, thank you so much for the very detailed answer. if all the red ones are really "up to rotate", this gives quite a lot of options/possibilities. I feel really bad to ask an other question, but I tried to get the simulator up to play around with that, but I don't get it into sleep mode - the "watch" is always awake. has anyone a hint for me, why this happens and how I can force it into sleep mode? It is not urgent, since I won't have time before the weekend anyway...

**Comment by wryun** [comment]
I'm a bit of a broken record on this one, but you can theoretically also do DMA transfers on a timer while the CPU is asleep ('Automated Bit Mapping'). If you use this, it should be possible to do arbitrary things with the display, though this will presumably use some more power.

One day I will actually try to do this :)

**Reply by joeycastillo** [reply]
One note for when you do: you’ll need to take the RAM out of back biased mode to do this (BBIASHS to 0, the relevant code should be in watch_init). The DMA controller can’t access RAM when back biased, though it seems to be only a difference of about 300 nA.

**Comment by aa-w** [comment]
Hi,
Apologies for dredging up an old thread but I was looking for some advice about whether you thought it would be possible to pull off a rotating animation through all of the segments without pulling the CPU out of sleep mode?
Would it just be a case of rewriting the call to slcd_sync_start_animation and specifying the number of segs I want it to rotate through?
Thanks in advance for the continued support of this project :)
